### ISSUE TITLE
Complete HPC Setup Guide: Building VLM System on SDSU Innovator Cluster

### ISSUE THREAD

**Issue by Mr-Parth24** [Mr-Parth24/VLM#2]
# Complete HPC Setup Guide: Building VLM System on SDSU Innovator Cluster

### Current Status ✅
- **Connected to:** cllogin002 (SDSU Innovator Cluster)
- **Username:** ppatel1447@jacks.local
- **Home Directory:** `/home/jacks.local/ppatel1447` (100GB quota, 0% used)
- **Scratch Directory:** `/scratch/jacks.local/ppatel1447` (empty, ready for large datasets)
- **Total Storage:** 60TB cluster storage with 52TB available

### 1.1 What is HPC?
**High-Performance Computing (HPC)** clusters are systems with:
- Multiple compute nodes (powerful computers)
- Shared storage systems
- Job scheduling systems (SLURM)
- Specialized software for parallel computing

### 1.2 Key HPC Concepts
- **Login Nodes:** Where you connect and prepare work (cllogin002)
- **Compute Nodes:** Where actual computation happens
- **Job Scheduler:** SLURM - manages and queues your computational jobs
- **Modules:** Pre-installed software packages you can load

### 1.3 File System Structure
```bash
/home/jacks.local/ppatel1447/          # Your home directory (100GB, permanent)
├── projects/                          # Your project codes
├── scripts/                          # Job submission scripts
└── environments/                     # Python virtual environments

/scratch/jacks.local/ppatel1447/       # Temporary storage (large datasets)
├── datasets/                         # SPIQA dataset
├── models/                          # Downloaded VLM models
└── results/                         # Processing outputs
```

### 2.1 Navigation and File Management
```bash
# Check current location
pwd

# List files and directories
ls -la                    # Detailed list
ls -lh                    # Human readable sizes

# Navigate directories
cd /scratch/jacks.local/ppatel1447    # Go to scratch
cd ~                      # Go to home directory
cd ..                     # Go up one directory

# Create directories
mkdir -p ~/projects/vlm-system        # Create project directory
mkdir -p /scratch/jacks.local/ppatel1447/{datasets,models,results}

# Check disk usage
df -h ~                   # Check home directory space
du -sh /scratch/jacks.local/ppatel1447/*  # Check scratch usage
```

### 2.2 File Transfer Commands
```bash
# Download files from internet
wget https://example.com/file.zip
curl -O https://example.com/file.zip

# Transfer files to/from your local computer
# From your local computer (PowerShell):
scp file.txt ppatel1447@jacks.local@innovator.sdstate.edu:~/projects/
scp -r folder/ ppatel1447@jacks.local@innovator.sdstate.edu:~/projects/

# From HPC to your computer:
scp ppatel1447@jacks.local@innovator.sdstate.edu:~/results/output.json ./
```

### 2.3 Text Editing
```bash
# Using nano (beginner-friendly)
nano script.py           # Create/edit Python script
# Ctrl+X to exit, Y to save

# Using vim (more advanced)
vim script.py
# Press 'i' to insert, Esc then ':wq' to save and quit

# View files without editing
cat file.txt             # Display entire file
less file.txt            # View large files (q to quit)
head -20 file.txt        # First 20 lines
tail -20 file.txt        # Last 20 lines
```

### 3.1 Understanding Modules
Modules are pre-installed software packages you can load into your environment.

```bash
# List available modules
module avail

# Search for specific software
module avail python      # Find Python versions
module avail cuda        # Find CUDA versions
module avail gcc         # Find compilers

# Load modules
module load python/3.9   # Load Python 3.9
module load cuda/11.8    # Load CUDA for GPU support
module load gcc/9.3.0    # Load GCC compiler

# List loaded modules
module list

# Unload modules
module unload python/3.9
module purge             # Unload all modules

# Get module information
module show python/3.9
```

### 3.2 Common Modules for VLM Project
```bash
# Essential modules for your project
module load python/3.9
module load cuda/11.8
module load gcc/9.3.0
module load git/2.34.1
```

### 3.3 Making Modules Persistent
```bash
# Create a module loading script
nano ~/.bashrc

# Add these lines to automatically load modules:
module load python/3.9
module load cuda/11.8
module load gcc/9.3.0
module load git/2.34.1

# Reload your shell
source ~/.bashrc
```

### 4.1 Creating Virtual Environment
```bash
# Load Python module
module load python/3.9

# Create virtual environment in home directory
cd ~
python -m venv environments/vlm-env

# Activate virtual environment
source environments/vlm-env/bin/activate

# Your prompt should now show (vlm-env)
# Deactivate when needed
deactivate
```

### 4.2 Installing Python Packages
```bash
# Activate your environment
source ~/environments/vlm-env/bin/activate

# Upgrade pip
pip install --upgrade pip

# Install basic packages
pip install numpy pandas matplotlib seaborn tqdm

# Install PyTorch with CUDA support
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

# Install Transformers and related libraries
pip install transformers accelerate bitsandbytes
pip install datasets pillow opencv-python-headless
pip install huggingface-hub

# Install additional packages for VLM models
pip install sentencepiece protobuf
pip install ipython jupyter  # For interactive development

# Save installed packages
pip freeze > requirements.txt
```

### 4.3 Testing GPU Access
```bash
# Create a GPU test script
nano test_gpu.py
```

Add this content:
```python
import torch
print(f"PyTorch version: {torch.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"CUDA version: {torch.version.cuda}")
    print(f"Number of GPUs: {torch.cuda.device_count()}")
    for i in range(torch.cuda.device_count()):
        print(f"GPU {i}: {torch.cuda.get_device_name(i)}")
        print(f"Memory: {torch.cuda.get_device_properties(i).total_memory / 1e9:.1f} GB")
```

```bash
# Run the test
python test_gpu.py
```

### 5.1 Understanding SLURM
SLURM (Simple Linux Utility for Resource Management) schedules and manages jobs on the cluster.

### 5.2 Basic SLURM Commands
```bash
# Check available partitions (queues)
sinfo

# Check job queue
squeue                   # All jobs
squeue -u ppatel1447     # Your jobs only

# Job information
scontrol show job JOBID  # Detailed job info
sacct                    # Job history

# Cancel jobs
scancel JOBID           # Cancel specific job
scancel -u ppatel1447   # Cancel all your jobs
```

### 5.3 Creating Job Scripts
Create a basic job script template:
```bash
nano ~/scripts/basic_job.sbatch
```

Add this content:
```bash
#!/bin/bash
#SBATCH --job-name=vlm-test          # Job name
#SBATCH --output=%j.out              # Output file (%j = job ID)
#SBATCH --error=%j.err               # Error file
#SBATCH --time=01:00:00              # Time limit (HH:MM:SS)
#SBATCH --partition=gpu              # Partition name
#SBATCH --gres=gpu:1                 # Request 1 GPU
#SBATCH --cpus-per-task=4            # Number of CPU cores
#SBATCH --mem=16G                    # Memory request

# Load modules
module load python/3.9
module load cuda/11.8

# Activate virtual environment
source ~/environments/vlm-env/bin/activate

# Change to scratch directory for large data
cd /scratch/jacks.local/ppatel1447

# Your commands here
echo "Job started at: $(date)"
echo "Running on node: $(hostname)"
python ~/test_gpu.py
echo "Job finished at: $(date)"
```

### 5.4 Submitting Jobs
```bash
# Submit a job
sbatch ~/scripts/basic_job.sbatch

# Check job status
squeue -u ppatel1447

# View job output
cat JOBID.out            # Replace JOBID with actual job ID
cat JOBID.err            # Check for errors
```

### 6.1 Creating Project Structure
```bash
# Create main project directory
mkdir -p ~/projects/vlm-system
cd ~/projects/vlm-system

# Create subdirectories
mkdir -p {src,config,scripts,notebooks,docs}
mkdir -p /scratch/jacks.local/ppatel1447/{datasets,models,results,temp}

# Create symbolic links to scratch for easy access
ln -s /scratch/jacks.local/ppatel1447/datasets ~/projects/vlm-system/datasets
ln -s /scratch/jacks.local/ppatel1447/models ~/projects/vlm-system/models
ln -s /scratch/jacks.local/ppatel1447/results ~/projects/vlm-system/results

# Initialize git repository
cd ~/projects/vlm-system
git init
git remote add origin https://github.com/Mr-Parth24/VLM.git
```

### 6.2 Create Configuration Files
```bash
# Create requirements file
nano ~/projects/vlm-system/requirements.txt
```

Add:
```txt
torch>=2.0.0
torchvision>=0.15.0
torchaudio>=2.0.0
transformers>=4.35.0
accelerate>=0.24.0
datasets>=2.14.0
pillow>=9.5.0
opencv-python-headless>=4.8.0
numpy>=1.24.0
pandas>=2.0.0
tqdm>=4.65.0
huggingface-hub>=0.17.0
sentencepiece>=0.1.99
protobuf>=4.24.0
bitsandbytes>=0.41.0
```

```bash
# Create module loading script
nano ~/projects/vlm-system/load_modules.sh
```

Add:
```bash
#!/bin/bash
module purge
module load python/3.9
module load cuda/11.8
module load gcc/9.3.0
module load git/2.34.1
echo "Modules loaded successfully!"
module list
```

```bash
# Make it executable
chmod +x ~/projects/vlm-system/load_modules.sh
```

### 7.1 Downloading Models
Create a model download script:
```bash
nano ~/projects/vlm-system/scripts/download_models.sbatch
```

Add:
```bash
#!/bin/bash
#SBATCH --job-name=download-models
#SBATCH --output=download_%j.out
#SBATCH --error=download_%j.err
#SBATCH --time=04:00:00
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=2
#SBATCH --mem=8G

# Load modules
source ~/projects/vlm-system/load_modules.sh

# Activate environment
source ~/environments/vlm-env/bin/activate

# Change to models directory
cd /scratch/jacks.local/ppatel1447/models

# Download models using huggingface-cli
echo "Downloading InternVL model..."
huggingface-cli download OpenGVLab/InternVL-Chat-V1-5 --local-dir ./internvl-chat-v1-5

echo "Downloading Qwen-VL model..."
huggingface-cli download Qwen/Qwen-VL-Chat --local-dir ./qwen-vl-chat

echo "Downloading mPLUG-Owl model..."
huggingface-cli download MAGAer13/mplug-owl2-llama2-7b --local-dir ./mplug-owl2-llama2-7b

echo "Model downloads completed!"
```

### 7.2 SPIQA Dataset Setup
```bash
nano ~/projects/vlm-system/scripts/download_spiqa.sbatch
```

Add:
```bash
#!/bin/bash
#SBATCH --job-name=download-spiqa
#SBATCH --output=spiqa_%j.out
#SBATCH --error=spiqa_%j.err
#SBATCH --time=02:00:00
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=2
#SBATCH --mem=16G

# Load modules and activate environment
source ~/projects/vlm-system/load_modules.sh
source ~/environments/vlm-env/bin/activate

# Change to datasets directory
cd /scratch/jacks.local/ppatel1447/datasets

# Clone SPIQA repository
git clone https://github.com/google/spiqa.git

# Download dataset files (adjust URLs as needed)
echo "Downloading SPIQA dataset..."
# Add specific download commands for SPIQA dataset here

echo "Dataset download completed!"
```

### 8.1 Setting Up Jupyter for Development
```bash
# Install Jupyter in your environment
source ~/environments/vlm-env/bin/activate
pip install jupyter jupyterlab

# Create Jupyter config
jupyter notebook --generate-config

# Create a Jupyter job script
nano ~/scripts/jupyter.sbatch
```

Add:
```bash
#!/bin/bash
#SBATCH --job-name=jupyter
#SBATCH --output=jupyter_%j.out
#SBATCH --error=jupyter_%j.err
#SBATCH --time=04:00:00
#SBATCH --partition=gpu
#SBATCH --gres=gpu:1
#SBATCH --cpus-per-task=4
#SBATCH --mem=32G

# Load modules and environment
source ~/projects/vlm-system/load_modules.sh
source ~/environments/vlm-env/bin/activate

# Get the node name
NODE=$(hostname)
PORT=8888

echo "Starting Jupyter Lab on node $NODE, port $PORT"
echo "SSH tunnel command:"
echo "ssh -N -L $PORT:$NODE:$PORT ppatel1447@jacks.local@innovator.sdstate.edu"

# Start Jupyter Lab
cd ~/projects/vlm-system
jupyter lab --no-browser --port=$PORT --ip=0.0.0.0
```

### 8.2 VS Code Setup (Optional)
If you prefer VS Code, you can use Remote SSH:
1. Install "Remote - SSH" extension in VS Code
2. Connect to: `ppatel1447@jacks.local@innovator.sdstate.edu`
3. Open folder: `/home/jacks.local/ppatel1447/projects/vlm-system`

### 9.1 Create Test Script
```bash
nano ~/projects/vlm-system/src/test_vlm.py
```

Add:
```python
import torch
from transformers import AutoModel, AutoTokenizer
from PIL import Image
import sys
import os

def test_internvl():
    """Test InternVL model"""
    print("Testing InternVL model...")
    
    # Check if model exists
    model_path = "/scratch/jacks.local/ppatel1447/models/internvl-chat-v1-5"
    if not os.path.exists(model_path):
        print(f"Model not found at {model_path}")
        return False
    
    try:
        # Load model and tokenizer
        model = AutoModel.from_pretrained(
            model_path,
            torch_dtype=torch.float16,
            device_map="auto",
            trust_remote_code=True
        )
        tokenizer = AutoTokenizer.from_pretrained(
            model_path,
            trust_remote_code=True
        )
        
        print("✅ InternVL model loaded successfully!")
        print(f"Model device: {next(model.parameters()).device}")
        return True
        
    except Exception as e:
        print(f"❌ Error loading InternVL: {e}")
        return False

def main():
    print("VLM Model Test")
    print("=" * 50)
    
    # Check GPU availability
    print(f"CUDA available: {torch.cuda.is_available()}")
    if torch.cuda.is_available():
        print(f"GPU: {torch.cuda.get_device_name(0)}")
        print(f"GPU Memory: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")
    
    print("\n" + "=" * 50)
    
    # Test models
    internvl_success = test_internvl()
    
    print("\n" + "=" * 50)
    print("Test Summary:")
    print(f"InternVL: {'✅ PASS' if internvl_success else '❌ FAIL'}")

if __name__ == "__main__":
    main()
```

### 9.2 Create Test Job Script
```bash
nano ~/projects/vlm-system/scripts/test_vlm.sbatch
```

Add:
```bash
#!/bin/bash
#SBATCH --job-name=test-vlm
#SBATCH --output=test_vlm_%j.out
#SBATCH --error=test_vlm_%j.err
#SBATCH --time=01:00:00
#SBATCH --partition=gpu
#SBATCH --gres=gpu:1
#SBATCH --cpus-per-task=4
#SBATCH --mem=32G

# Load modules and environment
source ~/projects/vlm-system/load_modules.sh
source ~/environments/vlm-env/bin/activate

# Run test
cd ~/projects/vlm-system
python src/test_vlm.py
```

### 9.3 Submit and Monitor
```bash
# Submit the test job
sbatch ~/projects/vlm-system/scripts/test_vlm.sbatch

# Monitor the job
squeue -u ppatel1447

# Check output when complete
cat test_vlm_JOBID.out
```

### 10.1 Login and Setup
```bash
# Connect to HPC
ssh ppatel1447@jacks.local@innovator.sdstate.edu

# Load your environment
source ~/projects/vlm-system/load_modules.sh
source ~/environments/vlm-env/bin/activate
cd ~/projects/vlm-system
```

### 10.2 Development Workflow
```bash
# Check job queue
squeue -u ppatel1447

# Submit a job
sbatch scripts/your_job.sbatch

# Check job output
tail -f JOBID.out      # Follow output in real-time

# Transfer files to/from local machine
# From your local computer:
scp local_file.py ppatel1447@jacks.local@innovator.sdstate.edu:~/projects/vlm-system/src/

# Download results to local:
scp ppatel1447@jacks.local@innovator.sdstate.edu:~/projects/vlm-system/results/* ./local_results/
```

### 10.3 Monitoring and Maintenance
```bash
# Check disk usage
df -h ~
du -sh /scratch/jacks.local/ppatel1447/*

# Clean up old job files
find ~ -name "*.out" -mtime +7 -delete    # Delete .out files older than 7 days
find ~ -name "*.err" -mtime +7 -delete    # Delete .err files older than 7 days

# Update packages
source ~/environments/vlm-env/bin/activate
pip install --upgrade pip
pip install --upgrade transformers torch
```

#### 1. "Permission Denied" Errors
```bash
# Check file permissions
ls -la filename
# Fix permissions
chmod +x script.sh
```

#### 2. "Module not found" Errors
```bash
# List available modules
module avail
# Load correct module
module load python/3.9
```

#### 3. "Out of Memory" Errors
- Increase memory in SBATCH script: `#SBATCH --mem=64G`
- Use gradient checkpointing in models
- Reduce batch size

#### 4. GPU Not Available
```bash
# Check GPU partitions
sinfo -p gpu
# Request GPU in job script
#SBATCH --gres=gpu:1
```

#### 5. Job Stuck in Queue
```bash
# Check partition limits
scontrol show partition gpu
# Use different partition or reduce resource requests
```

### HPC Learning
- [SLURM Documentation](https://slurm.schedmd.com/documentation.html)
- [HPC University](https://hpcuniversity.org/)
- [Linux Command Line Basics](https://linuxcommand.org/)

### VLM Development
- [Hugging Face Transformers](https://huggingface.co/docs/transformers/index)
- [PyTorch Documentation](https://pytorch.org/docs/stable/index.html)
- [CUDA Programming](https://docs.nvidia.com/cuda/)

### File Management
- [rsync Tutorial](https://www.digitalocean.com/community/tutorials/how-to-use-rsync-to-sync-local-and-remote-directories)
- [Git for Research](https://swcarpentry.github.io/git-novice/)

## 🎯 Next Steps

1. **Complete Environment Setup** (Days 1-3)
   - Master basic commands
   - Set up virtual environment
   - Test GPU access

2. **Download and Test Models** (Days 4-7)
   - Download VLM models
   - Download SPIQA dataset
   - Run initial tests

3. **Develop Processing Pipeline** (Days 8-14)
   - Implement data loading
   - Create VLM interfaces
   - Build processing scripts

4. **Scale Up Production** (Week 3+)
   - Optimize job scripts
   - Process full dataset
   - Analyze results

## ⚡ Quick Start Checklist

- [ ] Connect to HPC successfully
- [ ] Create virtual environment
- [ ] Install required packages
- [ ] Test GPU access
- [ ] Download one VLM model
- [ ] Submit first SLURM job
- [ ] Set up project structure
- [ ] Run VLM test script
- [ ] Process sample image
- [ ] Save results to JSON

Follow this guide step by step, and you'll have a fully functional VLM processing system on your HPC cluster!
